File: data/recomposed_projects/deepseek-coder-33b-instruct/signature/commons-graph/src/main/org/apache/commons/graph/model/DirectedMutableGraph.py
from __future__ import annotations
import re
import io
import typing
from typing import *
from src.main.org.apache.commons.graph.DirectedGraph import *
from src.main.org.apache.commons.graph.VertexPair import *
from src.main.org.apache.commons.graph.model.BaseMutableGraph import *


class DirectedMutableGraph(DirectedGraph, BaseMutableGraph):

    __outbound: typing.Dict[typing.Any, typing.Set[typing.Any]] = {}

    __inbound: typing.Dict[typing.Any, typing.Set[typing.Any]] = {}

    __serialVersionUID: int = 630111985439492792

    def _decorateRemoveVertex(self, v: typing.Any) -> None:

        if v in self.__inbound:
            del self.__inbound[v]

        if v in self.__outbound:
            del self.__outbound[v]

    def _decorateRemoveEdge(self, e: typing.Any) -> None:

        vertices = self.getVertices1(e)
        self.__inbound[vertices.getTail()].remove(vertices.getHead())
        self.__outbound[vertices.getHead()].remove(vertices.getTail())

    def _decorateAddVertex(self, v: typing.Any) -> None:
        self.__inbound[v] = set()
        self.__outbound[v] = set()

    def _decorateAddEdge(
        self, head: typing.Any, e: typing.Any, tail: typing.Any
    ) -> None:

        if tail in self.__inbound:
            self.__inbound[tail].add(head)
        else:
            self.__inbound[tail] = {head}

        if head in self.__outbound:
            self.__outbound[head].add(tail)
        else:
            self.__outbound[head] = {tail}

    def getOutDegree(self, v: typing.Any) -> int:
        return len(self.__outbound.get(v, set()))

    def getOutbound(self, v: typing.Any) -> typing.Iterable[typing.Any]:
        return self.__outbound.get(v)

    def getInDegree(self, v: typing.Any) -> int:
        return len(self.__inbound.get(v, set()))

    def getInbound(self, v: typing.Any) -> typing.Iterable[typing.Any]:
        return self.__inbound.get(v)

    def getDegree(self, v: typing.Any) -> int:
        return self.getInDegree(v) + self.getOutDegree(v)

File: java_projects/cleaned_final_projects_evosuite/commons-graph/src/main/java/org/apache/commons/graph/model/DirectedMutableGraph.java
package org.apache.commons.graph.model;


/*
  Licensed to the Apache Software Foundation (ASF) under one or more
  contributor license agreements.  See the NOTICE file distributed with
  this work for additional information regarding copyright ownership.
  The ASF licenses this file to You under the Apache License, Version 2.0
  (the "License"); you may not use this file except in compliance with
  the License.  You may obtain a copy of the License at

      http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.
 */


import org.apache.commons.graph.DirectedGraph;
import org.apache.commons.graph.VertexPair;

import java.util.HashMap;
import java.util.LinkedHashSet;
import java.util.Map;
import java.util.Set;

/**
 * A memory-based implementation of a mutable directed Graph.
 *
 * <p>This class is NOT thread safe!
 *
 * @param <V> the Graph vertices type
 * @param <E> the Graph edges type
 */
public class DirectedMutableGraph<V, E> extends BaseMutableGraph<V, E>
        implements DirectedGraph<V, E> {

    private static final long serialVersionUID = 630111985439492792L;

    private final Map<V, Set<V>> inbound = new HashMap<V, Set<V>>();

    private final Map<V, Set<V>> outbound = new HashMap<V, Set<V>>();

    /** {@inheritDoc} */
    @Override
    protected void decorateAddEdge(V head, E e, V tail) {
        inbound.get(tail).add(head);
        outbound.get(head).add(tail);
    }

    /** {@inheritDoc} */
    @Override
    protected void decorateAddVertex(V v) {
        inbound.put(v, new LinkedHashSet<V>());
        outbound.put(v, new LinkedHashSet<V>());
    }

    /** {@inheritDoc} */
    @Override
    protected void decorateRemoveEdge(E e) {
        final VertexPair<V> vertices = getVertices1(e);
        inbound.get(vertices.getTail()).remove(vertices.getHead());
        outbound.get(vertices.getHead()).remove(vertices.getTail());
    }

    /** {@inheritDoc} */
    @Override
    protected void decorateRemoveVertex(V v) {
        inbound.remove(v);
        outbound.remove(v);
    }

    /** {@inheritDoc} */
    public final int getDegree(V v) {
        return getInDegree(v) + getOutDegree(v);
    }

    /** {@inheritDoc} */
    public final Iterable<V> getInbound(V v) {
        return inbound.get(v);
    }

    /** {@inheritDoc} */
    public final int getInDegree(V v) {
        return inbound.get(v).size();
    }

    /** {@inheritDoc} */
    public final Iterable<V> getOutbound(V v) {
        return outbound.get(v);
    }

    /** {@inheritDoc} */
    public final int getOutDegree(V v) {
        return outbound.get(v).size();
    }
}
